internal/dkg/state_machine.go

```diff
@@ -689,69 +694,81 @@ func hasTimedOut(details *DBState) bool {
+	if currentState.State != Fresh {
+		err = validateReshareForRemainers(currentState, terms)
+	}
+
+	return err
+}
+
+func validateForAllDKGs(currentState *DBState, terms *drand.ProposalTerms) error {
 	if terms == nil {
 		return ErrMissingTerms
 	}
+
 	// it shouldn't really be possible for the wrong beaconID to make its way here, but better safe than sorry :)
 	if currentState.BeaconID != terms.BeaconID {
 		return ErrInvalidBeaconID
 	}
 
-	targetSch, err := crypto.SchemeFromName(terms.SchemeID)
+	sch, err := crypto.SchemeFromName(terms.SchemeID)
 	if err != nil {
 		return ErrInvalidScheme
 	}
 
-	for _, participant := range terms.Joining {
-		id, err := key.IdentityFromProto(participant, targetSch)
-		if err != nil {
-			return fmt.Errorf("%w, participant error: %s, expected: %s", key.ErrInvalidKeyScheme, err.Error(), targetSch.Name)
-		}
-		if err := id.ValidSignature(); err != nil {
-			return key.ErrInvalidKeyScheme
-		}
+	err = validateJoinerSignatures(terms, sch)
+	if err != nil {
+		return err
 	}
 
 	if terms.Timeout.AsTime().Before(time.Now()) {
 		return ErrTimeoutReached
 	}
 
-	if int(terms.Threshold) > len(terms.Joining)+len(terms.Remaining) {
+	nodeCount := len(terms.Joining) + len(terms.Remaining)
+	if int(terms.Threshold) > nodeCount {
 		return ErrThresholdHigherThanNodeCount
 	}
 
-	if err := validateEpoch(currentState, terms); err != nil {
-		return err
-	}
-
-	// some terms (such as genesis seed) get set during the first epoch
-	// additionally, we can't have remainers, `GenesisTime` == `TransitionTime`, amongst other things
-	if terms.Epoch == 1 {
-		return validateFirstEpoch(terms)
+	if nodeCount < int(currentState.Threshold) {
+		return ErrNodeCountTooLow
 	}
 
-	if util.Contains(terms.Joining, terms.Leader) {
-		return ErrLeaderCantJoinAfterFirstEpoch
-	}
-
-	if len(terms.Remaining) == 0 {
-		return ErrNoNodesRemaining
+	if int(terms.Threshold) < dkg.MinimumT(nodeCount) {
+		return ErrThresholdTooLow
 	}
 
-	// there's no theoretical reason the leader can't be leaving, but from a practical perspective
-	// it makes sense in case e.g. the DKG fails or aborts
-	if util.Contains(terms.Leaving, terms.Leader) || !util.Contains(terms.Remaining, terms.Leader) {
-		return ErrLeaderNotRemaining
-	}
+	return validateEpoch(currentState, terms)
+}
 
-	// nodes joining after the first epoch accept some things at face value
-	// nodes already in the network shouldn't accept e.g. a change of genesis time
-	if currentState.State != Fresh {
-		return validateReshare(currentState, terms)
+func validateJoinerSignatures(terms *drand.ProposalTerms, targetSch *crypto.Scheme) error {
+	for _, participant := range terms.Joining {
+		id, err := key.IdentityFromProto(participant, targetSch)
+		if err != nil {
+			return fmt.Errorf("%w, participant error: %s, expected: %s", key.ErrInvalidKeyScheme, err.Error(), targetSch.Name)
+		}
+		if err := id.ValidSignature(); err != nil {
+			return key.ErrInvalidKeyScheme
+		}
```

Comment:
Why don't we validate the signatures of node remaining? That seems like an attack vector for a malicious leader: sneak invalid signatures/public key pairs in there and bam you can take down the network after a resharing?

I guess a secondary problem here is that we "trust" signatures and keys coming from the same proposal, so as a malicious leader I could create fake keys and include valid signatures anyway. 

Not sure what kind of problems that would cause down the line, FWIW, I think we could even get rid of the long term public keys:  after we've joined we can use our secret share to sign messages and that's impossible to fake.

Reply:
checking if a node ought to be remaining (ie. it was in a previous epoch) compares the key and signatures; to that end, a leader modifying a signature would appear as a `Remainer` not present in the prior epoch and fail.
The only route for this being an attack vector I can see would be on the DKG from v1.5.9 -> v2, as technically the sigs aren't validated because nobody was ever a `Joiner`. That said, we have no new parties joining